modules/services/wob.nix

```
@@ -0,0 +1,75 @@
+{ config, lib, pkgs, ... }:
+
+let
+  inherit (lib) getExe mkEnableOption mkIf mkOption mkPackageOption optional;
+
+  cfg = config.services.wob;
+  settingsFormat = pkgs.formats.ini { };
+
+  configFile = settingsFormat.generate "wob.ini" cfg.settings;
+in {
+  meta.maintainers = with lib.maintainers; [ Scrumplex ];
+
+  options.services.wob = {
+    enable = mkEnableOption "wob";
+    package = mkPackageOption pkgs "wob" { };
+
+    settings = mkOption {
+      type = settingsFormat.type;
+      default = { };
+      example = {
+        "" = {
+          border_size = 10;
+          height = 50;
+        };
+        "output.foo".name = "DP-1";
+        "style.muted".background_color = "032cfc";
+      };
+      description = ''
+        Configuration written to {file}`$XDG_CONFIG_HOME/wob/wob.ini`.
+        See {manpage}`wob.ini(5)` for documentation.
+      '';
+    };
+
+    systemd = mkEnableOption "systemd service and socket for wob"
+      // mkOption { default = true; };
```

```suggestion
    systemd = mkEnableOption "systemd service and socket for wob"
      // { default = true; };
```

Also, when would one not run this as a service? Perhaps this option is unnecessary?

---

The upstream documentation mentions running this using an `exec` statement in a WM configuration. On the off chance that someone might want to do that, while still having access to `.settings` of this module, they can disable the systemd units.

---

Regarding the added `mkOption`: Apparently this is desirable to preserve attribute checking. I have adopted this after it was brought up in a previous review in nixpkgs: https://github.com/NixOS/nixpkgs/pull/245005#discussion_r1430540757

---

Interesting, I haven't seen that before and I'm not certain what the "attribute checking" refers to. But if it is better, then I don't mind 🙂

---

I did a quick test in repl:

```nix
nix-repl> lib.mkEnableOption "test"
{ _type = "option"; default = false; description = "Whether to enable test."; example = true; type = { ... }; }

nix-repl> lib.mkEnableOption "test" // { foo = "bar"; }
{ _type = "option"; default = false; description = "Whether to enable test."; example = true; foo = "bar"; type = { ... }; }

nix-repl> lib.mkEnableOption "test" // lib.mkOption { foo = "bar"; }
error:
       … in the right operand of the update (//) operator

         at «string»:1:27:

            1| lib.mkEnableOption "test" // lib.mkOption { foo = "bar"; }
             |                           ^

       … from call site

         at «string»:1:30:

            1| lib.mkEnableOption "test" // lib.mkOption { foo = "bar"; }
             |                              ^

       error: function 'mkOption' called with unexpected argument 'foo'

       at /nix/store/rqdf3gfjq8zh488msnv62h03kkzr308q-source/lib/options.nix:67:5:

           66|   mkOption =
           67|     {
             |     ^
           68|     # Default value used when no definition is given in the configuration.

```

So it basically prevents invalid attributes from getting into options

---

Very interesting, thanks for the investigation! It would have been nice if `mkEnableOption` had taken an additional attribute set parameter to allow such overrides in a more integrated way. I.e.

``` nix
lib.mkEnableOption "test" {
  default = true;
}
```

but I think it would be tricky to introduce now. It may be possible to have

``` nix
(lib.mkEnableOption "test").override {
  default = true;
}
```

though and maintain backwards compatibility. Not super pretty though.

In any case, the way you did it here with `mkOption` is fine.